### PR TITLE
DT-1971 retry dlq endpoint

### PR DESCRIPTION
**Pull request by mikehalmamoj**
Not quite ready for publishing these changes to the library yet but the PR is already too big.  Would also like to add security, telemetry and open api documentation if possible.

### CODE REVIEW

File: lib/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsQueueResource.kt [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+@RestController
+@RequestMapping("/queue-admin")
+class HmppsQueueResource(private val hmppsQueueService: HmppsQueueService) {
+
+  companion object {
+    private val log = LoggerFactory.getLogger(this::class.java)
+  }
+
+  @PutMapping("/retry-dlq/{dlqName}")
```
Comment (mikehalmamoj) [inline lib/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsQueueResource.kt:19]: Role based security coming soon. 

Not quite sure what to do about open api docs as we use both springfox and spring-docs.  Maybe some kind of `@Conditional` magic?

File: lib/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsQueue.kt [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+class HmppsQueue(
+  val sqsAwsClient: AmazonSQS,
+  val queueName: String,
+  val sqsAwsDlqClient: AmazonSQS,
+  val dlqName: String,
+) {
+  val queueUrl: String by lazy { sqsAwsClient.getQueueUrl(queueName).queueUrl }
+  val dlqUrl: String by lazy { sqsAwsDlqClient.getQueueUrl(dlqName).queueUrl }
+}
```
Comment (mikehalmamoj) [inline lib/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsQueue.kt:13]: This assumes all queues have a dlq which is always true afaik.  If we see a use case for a queue without a dlq then we can make the dlq nullable.

File: lib/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsQueueService.kt [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import org.springframework.stereotype.Service
+
+@Service
+class HmppsQueueService {
+
+  private val hmppsQueues: MutableList<HmppsQueue> = mutableListOf()
+
+  fun registerHmppsQueue(hmppsQueue: HmppsQueue) {
+    hmppsQueues += hmppsQueue
+  }
```
Comment (mikehalmamoj) [inline lib/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsQueueService.kt:16]: Slightly early optimisation to cater for multiple queues, but we know that is definitely a requirement so worth getting it in now.

File: lib/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsQueueService.kt [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      return RetryDlqResult(messageCount, messages.toList())
+    }
+}
+
+data class RetryDlqRequest(val hmppsQueue: HmppsQueue)
+data class RetryDlqResult(val messagesFoundCount: Int, val messages: List<Message>)
+
+private fun AmazonSQS.countMessagesOnQueue(queueUrl: String): Int =
+  this.getQueueAttributes(queueUrl, listOf("ApproximateNumberOfMessages"))
+    .let { it.attributes["ApproximateNumberOfMessages"]?.toInt() ?: 0 }
```
Comment (mikehalmamoj) [inline lib/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsQueueService.kt:42]: Probably break out utilities like this into another class or file.  Not quite sure whether we should ever make this public - doesn't feel right to extend a 3rd party API in a public way, but maybe in future :man_shrugging:

File: lib/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsQueueService.kt [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+      repeat(messageCount) {
+        sqsAwsDlqClient.receiveMessage(ReceiveMessageRequest(dlqUrl).withMaxNumberOfMessages(1)).messages.firstOrNull()
+          ?.let { msg ->
+            sqsAwsClient.sendMessage(queueUrl, msg.body)
+            sqsAwsDlqClient.deleteMessage(DeleteMessageRequest(dlqUrl, msg.receiptHandle))
+            messages += msg
+          }
+      }
+      return RetryDlqResult(messageCount, messages.toList())
+    }
```
Comment (mikehalmamoj) [inline lib/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsQueueService.kt:34]: Telemetry for this coming soon.

File: lib/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsQueueService.kt [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+data class RetryDlqRequest(val hmppsQueue: HmppsQueue)
+data class RetryDlqResult(val messagesFoundCount: Int, val messages: List<Message>)
+
+internal fun AmazonSQS.countMessagesOnQueue(queueUrl: String): Int =
+  this.getQueueAttributes(queueUrl, listOf("ApproximateNumberOfMessages"))
+    .let { it.attributes["ApproximateNumberOfMessages"]?.toInt() ?: 0 }
```
Comment (mikehalmamoj) [inline lib/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsQueueService.kt:42]: Probably break out utilities like this into another class or file. Not quite sure whether we should ever make this public - doesn't feel right to extend a 3rd party API in a public way, but maybe in future :man_shrugging:

File: lib/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsQueue.kt (hunk 2 above)
Comment (andymarke) [reply lib/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsQueue.kt:13]: I think court-register uses a queue to send a message, not receive - so no DLQ. But maybe they don't need this library

File: lib/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsQueueResource.kt (hunk 1 above)
Comment (andymarke) [reply lib/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsQueueResource.kt:19]: Maybe choose one and others can document manually???

**Review by andymarke — APPROVED** [review]
LGTM

File: lib/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsQueue.kt (hunk 2 above)
Comment (mikehalmamoj) [reply lib/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsQueue.kt:13]: Cool.  I think court-register will need this library when we start implementing the queue health indicators so we can tackle it then.

**Pull request merged**